### PR TITLE
chore: custom widget

### LINKED ISSUE #28607: [task]: [custom widgets] Drag and drop, search and property pane controls
Tasks:
- [ ]  Add the custom widget icon to the widget panel.
    - [ ]  Ensure that the custom widget icon is displayed when the search result is empty.
- [ ]  Enable drag-and-drop functionality for custom widgets and property pane controls.
    - [ ]  Implement the editing source control.
    - [ ]  Add control for managing events, including the ability to edit existing events.
    - [ ]  Include a help notice and link to provide guidance.
 
Designs - [figma](https://www.figma.com/file/dM8iA63UdQfVYMvO9u1eNB/%E2%9A%99%EF%B8%8F-Custom-Widget?type=design&node-id=194-6499&mode=design&t=OjElClAzOi7SFEao-0)
 
Expectations:
- [ ] The user should be able to view the custom widget icon on the widget panel
- [ ] The User should be able to view the custom widget icon if the search term doesn't match.
- [ ] The user should be able to see property controls for editing the source, Adding events, and help links.

### LINKED ISSUE #28608: [task]: [custom widgets] Appsmith custom widget SDK
Tasks:
- [ ]  Implement the Appsmith custom widget SDK with the following capabilities:
    - [ ]  Expose model, theme, and UI for customization.
        - [ ]  Implement the model, onModelChange, theme, onThemeChange, UI, and onUiChange features.
    - [ ]  expose render modes.
    - [ ]  Enable the triggering of custom events.
    - [ ]  Implement onReady event.
    - [ ]  Include support for import statements.
    - [ ]  CSS3 variables for UI and theme variables.

### LINKED ISSUE #28610: [task]: [custom widgets] Code mirror autocomplete, syntax highlight, formatting and linting

### LINKED ISSUE #28612: [task]: [custom widgets] Debugger drawer to view errors, logs logged from the custom widget code
Tasks:
- [ ]  Implement a debugger drawer to capture and display errors and logs for easier issue identification and debugging.
    - [ ]  Proxy to capture console function calls.
    - [ ]  onError event to catch all the unhandled events.
  
designs -[figma]( https://www.figma.com/file/dM8iA63UdQfVYMvO9u1eNB/%E2%9A%99%EF%B8%8F-Custom-Widget?type=design&node-id=166-11754&mode=design&t=tmxpDOrNT3jc7NQE-0)

### LINKED ISSUE #28615: [task]: [custom widgets] Default custom widget component for onboarding.
- [ ] Create a default component for the custom widget with helpful tips to guide new users.

### PR DESCRIPTION
**Pull request by sbalaji1192**
## Description
This PR's adds the custom widget.

#### PR fixes following issue(s)
Fixes #28607
Fixes #28610
Fixes #28615
Fixes #28608
Fixes #28612
> if no issue exists, please create an issue and ask the maintainers about this first
>
>
#### Type of change
- New feature (non-breaking change which adds functionality)
>
>
>
- [ ] Manual
- [ ] JUnit
- [x] Jest
- [x] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
  - Introduced search functionality with a wildcard option for unmatched widget searches.
  - Added new routes for custom widget editing.
  - Implemented additional editor modes for HTML and CSS.
  - Created new code templates and help components for custom widget builders.
  - Enhanced property controls with new button controls for custom widget editing.
  - Updated theming interfaces and constants for better theming support.
  - Added the CustomWidget component for embedding custom widgets in iframes.
  - Expanded widgets library to include the CustomWidget.

- **Enhancements**
  - Improved `CodeEditor` with additional modes, props, and resize behavior.
  - Enhanced `PropertyPaneControlConfig` interface with dynamic dependencies and additional properties.
  - Refined custom widget scripts with communication channel and event handling.

- **Bug Fixes**
  - Fixed visibility logic for the `ExternalWidget` card.
  
- **Documentation**
  - Added new messages and documentation links for custom widget features.

- **Tests**
  - Implemented new Cypress tests for custom widget default components and property pane interactions.
  - Updated workspace commands in Cypress tests.

- **Refactor**
  - Streamlined `PropertyControl` state management and editing functions.
  - Refactored code editor hint helper logic.
  
- **Style**
  - Added `borderLess` prop to style components without borders.

- **Chores**
  - Updated constants and messages related to custom widget features.
  - Adjusted webpack configuration to ignore specific module warnings.

### CODE REVIEW

File: app/client/src/components/propertyControls/CustomWidgetEditSourceButtonControl.tsx [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  render() {
+    return (
+      <Button
+        kind="secondary"
+        onClick={this.onCTAClick}
+        size="md"
+        style={{
+          width: "100%",
```
Comment (rajatagrawal) [inline app/client/src/components/propertyControls/CustomWidgetEditSourceButtonControl.tsx:167]: Is it okay to have inline styles ?

File: app/client/src/pages/Editor/CustomWidgetBuilder/Editor/CodeEditors/JSEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  CodeEditorBorder,
+  EditorModes,
+  EditorSize,
+  EditorTheme,
+  TabBehaviour,
+} from "components/editorComponents/CodeEditor/EditorConfig";
+import { CustomWidgetBuilderContext } from "../..";
+import LazyCodeEditor from "components/editorComponents/LazyCodeEditor";
+import { getAppsmithScriptSchema } from "widgets/CustomWidget/component/constants";
+// import { DebuggerLogType } from "../../types";
```
Comment (rajatagrawal) [inline app/client/src/pages/Editor/CustomWidgetBuilder/Editor/CodeEditors/JSEditor.tsx:14]: Commented out code

File: app/client/src/pages/Editor/CustomWidgetBuilder/Preview/Debugger/helpDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  return (
+    <Popover>
+      <PopoverTrigger>
+        <Button kind="tertiary" size="sm" startIcon="question-line" />
+      </PopoverTrigger>
+      <PopoverContent className={styles.consoleItemHelpContent}>
+        <List
+          items={[
+            // {
+            //   startIcon: <Icon name="book" size="md" />,
```
Comment (rajatagrawal) [inline app/client/src/pages/Editor/CustomWidgetBuilder/Preview/Debugger/helpDropdown.tsx:31]: commented out code

**Review by rajatagrawal — APPROVED** [review]

**Review by arunvjn — APPROVED** [review]
Code editor changes looks good to me

File: app/client/src/components/propertyControls/CustomWidgetEditSourceButtonControl.tsx (hunk 1 above)
Comment (sbalaji1192) [reply app/client/src/components/propertyControls/CustomWidgetEditSourceButtonControl.tsx:173]: removing it.

File: app/client/src/pages/Editor/CustomWidgetBuilder/Editor/CodeEditors/JSEditor.tsx (hunk 2 above)
Comment (sbalaji1192) [reply app/client/src/pages/Editor/CustomWidgetBuilder/Editor/CodeEditors/JSEditor.tsx:14]: We will be pushing an enhancement soon that will use this code.

File: app/client/src/pages/Editor/CustomWidgetBuilder/Preview/Debugger/helpDropdown.tsx (hunk 3 above)
Comment (sbalaji1192) [reply app/client/src/pages/Editor/CustomWidgetBuilder/Preview/Debugger/helpDropdown.tsx:31]: Same as above. We will need these options once we enabled them

**Review by rajatagrawal — APPROVED** [review]

**Review by rajatagrawal — APPROVED** [review]

**Pull request merged**